src/Strategy.sol

```diff
@@ -46,48 +108,138 @@ contract Strategy is BaseStrategy {
+            _addliquidity(_amountToDeposit);
+            _stakeLPTokens(unstakedLPBalance());
+        }
     }
 
     function liquidatePosition(uint256 _amountNeeded)
         internal
         override
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
-
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            unchecked {
-                _loss = _amountNeeded - totalAssets;
+        uint256 _liquidWant = wantBalance();
+
+        _amountNeeded = Math.min(_amountNeeded, estimatedTotalAssets()); // Otherwise we can end up declaring a liquidation loss when _amountNeeded is more than we own
+
+        if (_liquidWant < _amountNeeded) {
+            uint256 _lpTokensToSell = scaleWantToLP(_amountNeeded); // How many LP tokens do we need to get the required amount of `want`
+
+            uint256 _stakedLpTokens = stakedLPBalance(); // How many LP tokens do we have staked
+            uint256 _unstakedLPTokens = unstakedLPBalance(); // How many are available to unstake?
+            uint256 _requiredLPTokensToUnstake; // How many more LP tokens do we need to unstake to get the required amount of `want`.
+
+            // Free up the minimum amount of LP tokens to get to the amount of `want` we need
+            if (_unstakedLPTokens < _lpTokensToSell) {
+                _requiredLPTokensToUnstake =
+                    _lpTokensToSell -
+                    _unstakedLPTokens;
+                if (_stakedLpTokens >= _requiredLPTokensToUnstake) {
+                    _unstakeLPTokens(_requiredLPTokensToUnstake);
+                } else if (_stakedLpTokens > 0) {
+                    _unstakeLPTokens(_stakedLpTokens);
+                }
+            }
+
+            //withdraw from pool
+            uint256 _unstakedLpBalance = unstakedLPBalance();
+            if (_unstakedLpBalance > 0) {
+                _withdrawLiquidity(
+                    unstakedLPBalance(),
+                    _minAmountOfLPToWant(_unstakedLpBalance)
+                );
             }
+
+            _liquidWant = wantBalance();
+        }
+
+        if (_amountNeeded > _liquidWant) {
+            _liquidatedAmount = _liquidWant;
+            _loss = _amountNeeded - _liquidWant;
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
     function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
-        return want.balanceOf(address(this));
+        // unstake all staked token
+        synStakingMC.emergencyWithdraw(pid, address(this));
```

Comment:
Why do we use emergency withdraw here? It will cause us to lose rewards.

Reply:
Thought was, worse case scenario, we want to get our liquidity out without having to worry about rewards. Idea was that less things could go wrong if we do less while withdrawing our liquidity

Reply:
I see your comment below. I can change this to a normal withdraw and add a different function for emergency withdrawl